### PR TITLE
Detect Latest Package Revision

### PR DESCRIPTION
**Pull request by martinmaly**
The latest package revision (in semver order) is marked by a k8s resource tag
to make identifying it easier for clients.

### CODE REVIEW

File: porch/repository/pkg/cache/repository.go [hunk 1]
```diff
@@ -213,3 +246,60 @@ func (r *cachedRepository) pollOnce(ctx context.Context) {
 		klog.Warningf("error polling repo functions %s: %v", r.id, err)
 	}
 }
+
+func ToCachedPackageRevisionSlice(revisions []repository.PackageRevision) []*cachedPackageRevision {
```
Comment (justinsb) [inline porch/repository/pkg/cache/repository.go:250]: Nit: Do we need to export this?

File: porch/repository/pkg/cache/repository.go [hunk 2]
```diff
@@ -213,3 +246,60 @@ func (r *cachedRepository) pollOnce(ctx context.Context) {
...
+			PackageRevision:  revisions[i],
+			isLatestRevision: false,
+		}
+		result[i] = current
+	}
+	IdentifyLatestRevisions(result)
+	return result
+}
+
+func IdentifyLatestRevisions(result []*cachedPackageRevision) {
```
Comment (justinsb) [inline porch/repository/pkg/cache/repository.go:263]: Nit: Do we need to export this one also?

File: porch/repository/pkg/cache/repository.go [hunk 3]
```diff
@@ -213,3 +246,60 @@ func (r *cachedRepository) pollOnce(ctx context.Context) {
...
+	// Compute the latest among the different revisions of the same package.
+	// The map is keyed by the package name; Values are the latest revision found so far.
+	latest := map[string]*cachedPackageRevision{}
+	for _, current := range result {
+		current.isLatestRevision = false // Clear all values
+
+		// Check if the current package revision is more recent than the one sen so far.
```
Comment (justinsb) [inline porch/repository/pkg/cache/repository.go:270]: Typo: s/sen/seen

File: porch/repository/pkg/cache/repository.go [hunk 4]
```diff
@@ -213,3 +246,60 @@ func (r *cachedRepository) pollOnce(ctx context.Context) {
...
+		if current.Lifecycle() != v1alpha1.PackageRevisionLifecyclePublished {
+			continue
+		}
+
+		currentKey := current.Key()
+		if previous, ok := latest[currentKey.Package]; ok {
+			previousKey := previous.Key()
+			switch cmp := semver.Compare(currentKey.Revision, previousKey.Revision); {
+			case cmp == 0:
+				// Same revision.
```
Comment (justinsb) [inline porch/repository/pkg/cache/repository.go:281]: Perhaps warn?  Do we expect this?

**Review by justinsb — APPROVED** [review]
A few nits, but lgtm

File: porch/repository/pkg/cache/repository.go [hunk 5]
```diff
@@ -138,7 +165,9 @@ func (r *cachedRepository) CreatePackageRevision(ctx context.Context, obj *v1alp
 }
 
 func (r *cachedRepository) UpdatePackage(ctx context.Context, old repository.PackageRevision) (repository.PackageDraft, error) {
-	created, err := r.repo.UpdatePackage(ctx, old)
+	// Unwrap
+	unwrapped := old.(*cachedPackageRevision).PackageRevision
```
Comment (mortent) [inline porch/repository/pkg/cache/repository.go:169]: Why do we need to unwrap here? It seems like the `UpdatePackage` accepts a parameter of type `PackageRevision` which the `*cachedPackageRevision` implements.

**Review by mortent — APPROVED** [review]
Looks good to me. I just had one question.

File: porch/repository/pkg/cache/repository.go (hunk 1 above)
Comment (martinmaly) [reply porch/repository/pkg/cache/repository.go:250]: no; I'll hide it. good catch!

File: porch/repository/pkg/cache/repository.go (hunk 2 above)
Comment (martinmaly) [reply porch/repository/pkg/cache/repository.go:263]: ditto. will hide

File: porch/repository/pkg/cache/repository.go (hunk 3 above)
Comment (martinmaly) [reply porch/repository/pkg/cache/repository.go:270]: thanks!

File: porch/repository/pkg/cache/repository.go (hunk 4 above)
Comment (martinmaly) [reply porch/repository/pkg/cache/repository.go:281]: Warning seems like a good idea.

File: porch/repository/pkg/cache/repository.go (hunk 5 above)
Comment (martinmaly) [reply porch/repository/pkg/cache/repository.go:169]: the Git repository will cast the interface to *gitPackageRevision.

We could expand the interface and avoid the cast but Git is getting at some fields not exposed via the interface.

File: porch/repository/pkg/cache/repository.go [hunk 6]
```diff
@@ -86,14 +113,14 @@ func (r *cachedRepository) getPackages(ctx context.Context, forceRefresh bool) (
...
+		packages = ToCachedPackageRevisionSlice(p)
 
 		r.mutex.Lock()
-		r.cachedPackages = p
+		r.cachedPackages = packages
 		r.mutex.Unlock()
 	}
 
-	return packages, nil
+	return ToPackageRevisionSlice(packages), nil
```
[the remark this answers is not in the document]
Comment (martinmaly) [reply porch/repository/pkg/cache/repository.go:123]: Maybe the names are not ideal. The cache now stores `cachedPackages []*cachedPackageRevision` so on the 'way out' it needs to be converted to `[]PackageRevision`. So this call really just "changes the type of the slice".
When we store the packages, we wrap then and compute the latest one; then we return them, still wrapped, but re-typed to []PackageRevision.

We _could_ store the cache as []PackageRevision and perform casts; I thought this was more explicit given all the other casting that is happening.

**Pull request merged**